flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/handler/PulsarAdminInvocationHandler.java

```diff
@@ -0,0 +1,152 @@
+        Class<?> returnType = method.getReturnType();
+
+        // No need to proxy the void return type.
+        // The non-interface type is not able to proxy.
+        if (returnType.equals(Void.TYPE) || !returnType.isInterface()) {
+            return method.invoke(admin, args);
+        }
+
+        String methodName = method.getName();
+        if (handlers.containsKey(methodName)) {
+            return handlers.get(methodName);
+        }
+
+        Object handler =
+                Proxy.newProxyInstance(
+                        Thread.currentThread().getContextClassLoader(),
+                        new Class[] {returnType},
+                        new RequestHandler(method.invoke(admin, args)));
+        this.handlers.put(methodName, handler);
+
+        return handler;
+    }
+
+    /** A proxy handler with retry support for all the admin request. */
+    @SuppressWarnings({"java:S1193", "java:S1181", "java:S3776", "java:S112"})
+    private class RequestHandler implements InvocationHandler {
+
+        private final Object handler;
+
+        public RequestHandler(Object handler) {
+            this.handler = handler;
+        }
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return doInvoke(method, args, retryTimes);
+        }
+
+        private Object doInvoke(Method method, Object[] args, int remainingTimes) throws Throwable {
+            while (true) {
+                // Make sure the request is allowed in the given rates.
+                requestRateLimit(requestRates);
+
+                try {
+                    return method.invoke(handler, args);
+                } catch (InvocationTargetException e) {
+                    Optional<PulsarAdminException> optional =
+                            findThrowable(e, PulsarAdminException.class);
+                    if (optional.isPresent()) {
+                        PulsarAdminException ex = optional.get();
+                        if (ex instanceof NotFoundException) {
+                            // No need to retry on such exceptions.
+                            throw ex;
+                        }
+
+                        remainingTimes--;
+                        LOG.warn("Request error in Admin API, remain times: {}", remainingTimes, e);
+                        if (remainingTimes == 0) {
+                            throw ex;
+                        } else {
+                            // Sleep for the given times before executing the next query.
+                            sleepUninterruptibly(waitMillis, MILLISECONDS);
+                        }
+                    } else {
+                        throw e.getCause();
+                    }
+                }
+            }
+        }
+    }
+
+    /** Request a global ratelimit which limits the total admin API request rates. */
+    private static void requestRateLimit(int requestRates) {
+        if (rateLimiter == null) {
```

Comment:
Maybe initiate the `rateLimit` in a hungry style is enough. This can avoid the overhead of accessing volatile variable.

Reply:
`volatile` can fix some race condition in double lock syncing. I think.

Reply:
FYI. https://www.baeldung.com/java-singleton-double-checked-locking

Reply:
Think about it and find non-lazy singleton is not very suitable for this scene, keep this implementation is good.